lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -250,31 +250,46 @@ class MongoClientInterface {
     }
 
     getBucketAndObject(bucketName, objName, params, log, cb) {
-        this.getBucketAttributes(bucketName, log, (err, bucket) => {
-            if (err) {
-                log.error(
-                    'getBucketAttributes: error getting bucket attributes',
-                    { error: err.message });
-                return cb(err);
-            }
-            this.getObject(bucketName, objName, params, log, (err, obj) => {
-                if (err) {
-                    if (err === errors.NoSuchKey) {
-                        return cb(null,
-                                  { bucket:
-                                    BucketInfo.fromObj(bucket).serialize(),
-                                  });
+        return async.parallel({
+            bucket: callback => {
+                this.getBucketAttributes(bucketName, log, (err, bucket) => {
+                    if (err) {
+                        log.error(
+                            'getBucketAndObject: error getting bucket attrs',
+                            { error: err.message });
+                        return callback(err);
                     }
-                    log.error('getObject: error getting object',
-                    { error: err.message });
-                    return cb(err);
-                }
-                return cb(null, {
-                    bucket: BucketInfo.fromObj(bucket).serialize(),
-                    obj: JSON.stringify(obj),
+                    return callback(null, bucket);
+                });
+            },
+            obj: callback => {
+                this.getObject(bucketName, objName, params, log, (err, obj) => {
+                    if (err) {
+                        if (err === errors.NoSuchKey) {
+                            return callback(null, null);
+                        }
+                        log.error('getBucketAndObject: error getting object',
+                                  { error: err.message });
+                        return callback(err);
+                    }
+                    return callback(null, obj);
                 });
+            },
+        }, (err, results) => {
+            if (err) {
+                return cb(errors.InternalError);
+            }
+            const { bucket, obj } = results;
+            if (!obj) {
+                return cb(null,
+                          { bucket:
+                            BucketInfo.fromObj(bucket).serialize(),
```

Comment:
`BucketInfo.fromObj(bucket).serialize()` could be done in the async.parallel worker to avoid duplication since it's used in all of the successful code paths.